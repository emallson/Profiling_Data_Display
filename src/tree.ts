export type FrameTree<T> = {
  name: string;
  value: T;
  children: {
    [name: string]: FrameTree<T>;
  };
};

type FrameKey = {
  /**
   * The reversed frame path
   */
  framePath: string[];
  script?: string;
};

export function parseKey(key: string): FrameKey {
  const [rawFramePath, script] = key.split(":");

  return {
    framePath: rawFramePath.split("/"),
    script,
  };
}

export function punch<T>(
  tree: FrameTree<T | undefined>,
  path: string[]
): FrameTree<T | undefined> {
  if (path.length === 0) {
    return tree;
  }

  const key = path[path.length - 1];
  let child = tree.children[key];

  if (child === undefined) {
    child = {
      name: key,
      value: undefined,
      children: {},
    };
    tree.children[key] = child;
  }

  return punch(child, path.slice(0, -1));
}

export function build<T>(
  data: T[],
  selector: (value: T) => string
): FrameTree<T | undefined> {
  let root = {
    name: "root",
    value: undefined,
    children: {},
  };

  for (const element of data) {
    const { framePath, script } = parseKey(selector(element));
    const node = punch<T>(root, script ? [script, ...framePath] : framePath);
    node.value = element;
  }

  return root;
}
