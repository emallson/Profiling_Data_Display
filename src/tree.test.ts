import * as tree from "./tree";

describe("parseKey", () => {
  it("should correctly parse the frame path for a scriptless key", () => {
    const key =
      "RavenButton4TipFrame/RavenButton4Container/RavenButton4Frame/RavenBarGroupDebuffs/UIParent";

    expect(tree.parseKey(key)).toStrictEqual({
      framePath: [
        "RavenButton4TipFrame",
        "RavenButton4Container",
        "RavenButton4Frame",
        "RavenBarGroupDebuffs",
        "UIParent",
      ],
      script: undefined,
    });
  });

  it("should correctly parse a frame path with a script key", () => {
    const key =
      "RavenButton4TipFrame/RavenButton4Container/RavenButton4Frame/RavenBarGroupDebuffs/UIParent:OnLeave";

    expect(tree.parseKey(key)).toStrictEqual({
      framePath: [
        "RavenButton4TipFrame",
        "RavenButton4Container",
        "RavenButton4Frame",
        "RavenBarGroupDebuffs",
        "UIParent",
      ],
      script: "OnLeave",
    });
  });

  it("should correctly handle a singleton path", () => {
    const key = "nil";

    expect(tree.parseKey(key)).toStrictEqual({
      framePath: ["nil"],
      script: undefined,
    });
  });
});

describe("punch", () => {
  it("should create a new nodes recursively as needed", () => {
    const root: tree.FrameTree<undefined> = {
      name: "root",
      value: undefined,
      children: {},
    };

    const test1 = tree.punch(root, ["test1", "parent"]);
    const test2 = tree.punch(root, ["test2", "test1", "parent"]);
    const test1_copy = tree.punch(root, ["test1", "parent"]);

    expect(test1).toBe(test1_copy);
    expect(test1).toBe(root.children["parent"].children["test1"]);
    expect(test1).toBeDefined();
    expect(test2).toBeDefined();
  });
});

describe("buildTree", () => {
  it("should build a toy tree successfully", () => {
    const keys = [
      "WeakAurasFrame.24f4f0164d0.24f4f01a9e0/WeakAurasFrame.24f4f0164d0/WeakAurasFrame/UIParent:OnSizeChanged",
      "WeakAurasFrame.24ef3d108a0.24ef3d121f0/WeakAurasFrame.24ef3d108a0/WeakAurasFrame/UIParent:OnSizeChanged",
      "BattleGroundEnemies.Allies.24ef2416520.24ef242a7b0/BattleGroundEnemies.Allies.24ef2416520/BattleGroundEnemies.Allies/BattleGroundEnemies:OnHide",
      "BigWigsProximityAnchor.24f59eea100/BigWigsProximityAnchor/UIParent:OnEnter",
      "WeakAurasFrame.24ef9e33540.24ef9e369c0/WeakAurasFrame.24ef9e33540/WeakAurasFrame/UIParent:OnSizeChanged",
      "WeakAurasFrame.24f4f039dd0/WeakAurasFrame/UIParent:OnSizeChanged",
    ];

    const result = tree.build<string>(keys, (x) => x);

    expect(result).toMatchSnapshot();
  });
});
