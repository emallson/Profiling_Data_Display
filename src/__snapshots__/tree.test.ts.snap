// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTree should build a toy tree successfully 1`] = `
Object {
  "children": Object {
    "BattleGroundEnemies": Object {
      "children": Object {
        "BattleGroundEnemies.Allies": Object {
          "children": Object {
            "BattleGroundEnemies.Allies.24ef2416520": Object {
              "children": Object {
                "BattleGroundEnemies.Allies.24ef2416520.24ef242a7b0": Object {
                  "children": Object {
                    "OnHide": Object {
                      "children": Object {},
                      "name": "OnHide",
                      "value": "BattleGroundEnemies.Allies.24ef2416520.24ef242a7b0/BattleGroundEnemies.Allies.24ef2416520/BattleGroundEnemies.Allies/BattleGroundEnemies:OnHide",
                    },
                  },
                  "name": "BattleGroundEnemies.Allies.24ef2416520.24ef242a7b0",
                  "value": undefined,
                },
              },
              "name": "BattleGroundEnemies.Allies.24ef2416520",
              "value": undefined,
            },
          },
          "name": "BattleGroundEnemies.Allies",
          "value": undefined,
        },
      },
      "name": "BattleGroundEnemies",
      "value": undefined,
    },
    "UIParent": Object {
      "children": Object {
        "BigWigsProximityAnchor": Object {
          "children": Object {
            "BigWigsProximityAnchor.24f59eea100": Object {
              "children": Object {
                "OnEnter": Object {
                  "children": Object {},
                  "name": "OnEnter",
                  "value": "BigWigsProximityAnchor.24f59eea100/BigWigsProximityAnchor/UIParent:OnEnter",
                },
              },
              "name": "BigWigsProximityAnchor.24f59eea100",
              "value": undefined,
            },
          },
          "name": "BigWigsProximityAnchor",
          "value": undefined,
        },
        "WeakAurasFrame": Object {
          "children": Object {
            "WeakAurasFrame.24ef3d108a0": Object {
              "children": Object {
                "WeakAurasFrame.24ef3d108a0.24ef3d121f0": Object {
                  "children": Object {
                    "OnSizeChanged": Object {
                      "children": Object {},
                      "name": "OnSizeChanged",
                      "value": "WeakAurasFrame.24ef3d108a0.24ef3d121f0/WeakAurasFrame.24ef3d108a0/WeakAurasFrame/UIParent:OnSizeChanged",
                    },
                  },
                  "name": "WeakAurasFrame.24ef3d108a0.24ef3d121f0",
                  "value": undefined,
                },
              },
              "name": "WeakAurasFrame.24ef3d108a0",
              "value": undefined,
            },
            "WeakAurasFrame.24ef9e33540": Object {
              "children": Object {
                "WeakAurasFrame.24ef9e33540.24ef9e369c0": Object {
                  "children": Object {
                    "OnSizeChanged": Object {
                      "children": Object {},
                      "name": "OnSizeChanged",
                      "value": "WeakAurasFrame.24ef9e33540.24ef9e369c0/WeakAurasFrame.24ef9e33540/WeakAurasFrame/UIParent:OnSizeChanged",
                    },
                  },
                  "name": "WeakAurasFrame.24ef9e33540.24ef9e369c0",
                  "value": undefined,
                },
              },
              "name": "WeakAurasFrame.24ef9e33540",
              "value": undefined,
            },
            "WeakAurasFrame.24f4f0164d0": Object {
              "children": Object {
                "WeakAurasFrame.24f4f0164d0.24f4f01a9e0": Object {
                  "children": Object {
                    "OnSizeChanged": Object {
                      "children": Object {},
                      "name": "OnSizeChanged",
                      "value": "WeakAurasFrame.24f4f0164d0.24f4f01a9e0/WeakAurasFrame.24f4f0164d0/WeakAurasFrame/UIParent:OnSizeChanged",
                    },
                  },
                  "name": "WeakAurasFrame.24f4f0164d0.24f4f01a9e0",
                  "value": undefined,
                },
              },
              "name": "WeakAurasFrame.24f4f0164d0",
              "value": undefined,
            },
            "WeakAurasFrame.24f4f039dd0": Object {
              "children": Object {
                "OnSizeChanged": Object {
                  "children": Object {},
                  "name": "OnSizeChanged",
                  "value": "WeakAurasFrame.24f4f039dd0/WeakAurasFrame/UIParent:OnSizeChanged",
                },
              },
              "name": "WeakAurasFrame.24f4f039dd0",
              "value": undefined,
            },
          },
          "name": "WeakAurasFrame",
          "value": undefined,
        },
      },
      "name": "UIParent",
      "value": undefined,
    },
  },
  "name": "root",
  "value": undefined,
}
`;
